File: generated_code/cpp/test_unittest_2.cpp
#include <gtest/gtest.h>
#include <vector>
#include "solution_proxy.h"

// Helper function to create a linked list from a vector
ListNode* createList(const std::vector<int>& nums) {
    if (nums.empty()) return nullptr;
    ListNode* head = new ListNode(nums[0]);
    ListNode* curr = head;
    for (size_t i = 1; i < nums.size(); ++i) {
        curr->next = new ListNode(nums[i]);
        curr = curr->next;
    }
    return head;
}

// Helper function to convert a linked list back to a vector for verification
std::vector<int> listToVector(ListNode* head) {
    std::vector<int> result;
    ListNode* curr = head;
    while (curr != nullptr) {
        result.push_back(curr->val);
        curr = curr->next;
    }
    return result;
}

// Helper function to free the allocated linked list memory
void freeList(ListNode* head) {
    while (head != nullptr) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

class AddTwoNumbersTest : public ::testing::Test {
protected:
    Solution solution;
};

// Example 1: l1 = [2,4,3], l2 = [5,6,4] -> Output: [7,0,8]
TEST_F(AddTwoNumbersTest, TestExample1) {
    ListNode* l1 = createList({2, 4, 3});
    ListNode* l2 = createList({5, 6, 4});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {7, 0, 8};
    
    EXPECT_EQ(listToVector(result), expected);
    
    freeList(l1);
    freeList(l2);
    freeList(result);
}

// Example 2: l1 = [0], l2 = [0] -> Output: [0]
TEST_F(AddTwoNumbersTest, TestExample2SingleZeros) {
    ListNode* l1 = createList({0});
    ListNode* l2 = createList({0});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {0};
    
    EXPECT_EQ(listToVector(result), expected);
    
    freeList(l1);
    freeList(l2);
    freeList(result);
}

// Example 3: l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9] -> Output: [8,9,9,9,0,0,0,1]
TEST_F(AddTwoNumbersTest, TestExample3VaryingLengthsWithCarries) {
    ListNode* l1 = createList({9, 9, 9, 9, 9, 9, 9});
    ListNode* l2 = createList({9, 9, 9, 9});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {8, 9, 9, 9, 0, 0, 0, 1};
    
    EXPECT_EQ(listToVector(result), expected);
    
    freeList(l1);
    freeList(l2);
    freeList(result);
}

// Edge Case: One list is significantly longer than the other, ending with additional carry
TEST_F(AddTwoNumbersTest, TestOneListLongerWithFinalCarry) {
    ListNode* l1 = createList({1});
    ListNode* l2 = createList({9, 9, 9});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {0, 0, 0, 1};
    
    EXPECT_EQ(listToVector(result), expected);
    
    freeList(l1);
    freeList(l2);
    freeList(result);
}

// Edge Case: Maximum constraint length simulation (100 nodes of 9s)
TEST_F(AddTwoNumbersTest, TestMaxConstraintLength) {
    std::vector<int> longNum(100, 9);
    ListNode* l1 = createList(longNum);
    ListNode* l2 = createList({1});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    
    std::vector<int> expected(100, 0);
    expected.push_back(1); // 1 followed by 100 zeros in reverse order
    
    EXPECT_EQ(listToVector(result), expected);
    
    freeList(l1);
    freeList(l2);
    freeList(result);
}

// Edge Case: Minimum constraint length (1 node each, no carry)
TEST_F(AddTwoNumbersTest, TestMinConstraintLengthNoCarry) {
    ListNode* l1 = createList({5});
    ListNode* l2 = createList({3});
    
    ListNode* result = solution.addTwoNumbers(l1, l2);
    std::vector<int> expected = {8};
    
    EXPECT_EQ(listToVector(result), expected);
    
    freeList(l1);
    freeList(l2);
    freeList(result);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: generated_code/py/test_unittest_2.py
import unittest
from typing import List, Optional
from program_2_1 import Solution

# Definition for singly-linked list (needed for test setup and verification)
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class TestAddTwoNumbers(unittest.TestCase):

    def helper_create_linked_list(self, arr: List[int]) -> Optional[ListNode]:
        """Helper method to convert a Python list into a ListNode linked list."""
        if not arr:
            return None
        dummy = ListNode(0)
        current = dummy
        for val in arr:
            current.next = ListNode(val)
            current = current.next
        return dummy.next

    def helper_linked_list_to_list(self, head: Optional[ListNode]) -> List[int]:
        """Helper method to convert a ListNode linked list back into a Python list."""
        arr = []
        current = head
        while current:
            arr.append(current.val)
            current = current.next
        return arr

    def assert_linked_lists_equal(self, expected_arr: List[int], actual_head: Optional[ListNode]):
        """Custom assertion to verify the result linked list matches the expected array."""
        actual_arr = self.helper_linked_list_to_list(actual_head)
        self.assertEqual(actual_arr, expected_arr, f"Expected {expected_arr}, but got {actual_arr}")

    def test_example_1(self):
        """Test Example 1: l1 = [2,4,3], l2 = [5,6,4] -> Output: [7,0,8]"""
        sol = Solution()
        l1 = self.helper_create_linked_list([2, 4, 3])
        l2 = self.helper_create_linked_list([5, 6, 4])
        result = sol.addTwoNumbers(l1, l2)
        self.assert_linked_lists_equal([7, 0, 8], result)

    def test_example_2(self):
        """Test Example 2: l1 = [0], l2 = [0] -> Output: [0]"""
        sol = Solution()
        l1 = self.helper_create_linked_list([0])
        l2 = self.helper_create_linked_list([0])
        result = sol.addTwoNumbers(l1, l2)
        self.assert_linked_lists_equal([0], result)

    def test_example_3(self):
        """Test Example 3: l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9] -> Output: [8,9,9,9,0,0,0,1]"""
        sol = Solution()
        l1 = self.helper_create_linked_list([9, 9, 9, 9, 9, 9, 9])
        l2 = self.helper_create_linked_list([9, 9, 9, 9])
        result = sol.addTwoNumbers(l1, l2)
        self.assert_linked_lists_equal([8, 9, 9, 9, 0, 0, 0, 1], result)

    def test_different_lengths_l1_shorter(self):
        """Test lists of different lengths where l1 is shorter than l2."""
        sol = Solution()
        l1 = self.helper_create_linked_list([1])
        l2 = self.helper_create_linked_list([9, 9])
        result = sol.addTwoNumbers(l1, l2)
        self.assert_linked_lists_equal([0, 0, 1], result)

    def test_carry_creates_new_most_significant_digit(self):
        """Test where the final addition creates a carry that extends the list length."""
        sol = Solution()
        l1 = self.helper_create_linked_list([5])
        l2 = self.helper_create_linked_list([5])
        result = sol.addTwoNumbers(l1, l2)
        self.assert_linked_lists_equal([0, 1], result)

    def test_minimum_constraint_single_nodes(self):
        """Test minimum constraint values: Single nodes with low values without carry."""
        sol = Solution()
        l1 = self.helper_create_linked_list([1])
        l2 = self.helper_create_linked_list([2])
        result = sol.addTwoNumbers(l1, l2)
        self.assert_linked_lists_equal([3], result)

    def test_large_value_constraint_100_nodes(self):
        """Test maximum constraint boundaries: 100 nodes filled with 9s (triggering continuous carries)."""
        sol = Solution()
        l1 = self.helper_create_linked_list([9] * 100)
        l2 = self.helper_create_linked_list([9] * 100)
        result = sol.addTwoNumbers(l1, l2)
        
        # Adding 99...9 (100 digits) + 99...9 (100 digits) results in:
        # [8, 9, 9, ..., 9, 1] -> 18 followed by ninety-nine 9s, then 1 (total 101 digits)
        expected = [8] + ([9] * 99) + [1]
        self.assert_linked_lists_equal(expected, result)

if __name__ == '__main__':
    unittest.main()